File: 2707-extra-characters-in-a-string/2707-extra-characters-in-a-string.cpp
class Solution {
public:
    int minExtraChar(string s, vector<string>& dictionary) {
        unordered_map<string,int>mp;
        for(int i=0;i<dictionary.size();i++)
        {
            mp[dictionary[i]]=1;
        }
        vector<int>dp(s.size()+1,INT_MAX);
        dp[0]=0;
        for(int i=0;i<=s.size();i++)
        {
            string temp;
            int counter=dp[i];
            // cout<<dp[i]<<" ";
            for(int j=i+1;j<=s.size();j++)
            {
                temp.push_back(s[j-1]);
                counter++;
                if(mp[temp])
                {
                    dp[j]=min(dp[j],dp[i]);
                }
                else
                    dp[j]=min(dp[j],counter);
            }
        }
        return dp[s.size()];
        
    }
};

File: 2707-extra-characters-in-a-string/2707-extra-characters-in-a-string.java
class Solution {
    public int minExtraChar(String s, String[] dictionary) {
        HashMap<String,Integer> wordsAvailable = new HashMap<String,Integer>();
        for(String words : dictionary){
            wordsAvailable.put(words,1);
        }
        int [] minLeftlaoneWords = new int [s.length()+1];
        for(int i=1 ;i<=s.length();i++){
            minLeftlaoneWords[i]= Integer.MAX_VALUE;
            String word ="";
            for(int j=i ;j>0 ;j--){
                word = s.charAt(j-1)+ word;
                minLeftlaoneWords[i] = Math.min(minLeftlaoneWords[j-1]+wordsAvailable.getOrDefault(word,word.length()+1)-1 ,minLeftlaoneWords[i]);
            }
            // System.out.println(minLeftlaoneWords[i]);
        }
        return  minLeftlaoneWords[s.length()];
    }
}